core/document/src/lib.rs

```diff
@@ -22,31 +26,148 @@ impl SvgElement {
 	}
 }
 
-#[derive(Default, Debug, Clone, PartialEq)]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DocumentError {
+	ElementNotFound,
+	NotAShape,
+	ElementAlreadyExists,
+	InvalidPath,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Folder {
+	elements: HashMap<String, SvgElement>,
+}
+
+impl Folder {
+	fn add_element(&mut self, name: String, svg: SvgElement) -> Result<(), DocumentError> {
+		#[allow(clippy::map_entry)]
+		if self.elements.contains_key(&name) {
+			Err(DocumentError::ElementAlreadyExists)
+		} else {
+			self.elements.insert(name, svg);
+			Ok(())
+		}
+	}
+
+	fn remove_element(&mut self, name: &str) -> Result<(), DocumentError> {
+		self.elements.remove(name).map_or(Err(DocumentError::ElementNotFound), |_| Ok(()))
+	}
+
+	/// Returns a list of elements in the folder
+	pub fn list(&self) -> Keys<String, SvgElement> {
+		self.elements.keys()
+	}
+
+	fn element(&self, name: &str) -> Option<&SvgElement> {
+		self.elements.get(name)
+	}
+
+	fn mut_element(&mut self, name: &str) -> Option<&mut SvgElement> {
+		self.elements.get_mut(name)
+	}
+}
+
+impl Default for Folder {
+	fn default() -> Self {
+		Self { elements: HashMap::new() }
+	}
+}
+
+#[derive(Debug, Clone, PartialEq)]
 pub struct Document {
-	pub svg: Vec<SvgElement>,
+	pub svg: SvgElement,
+}
+
+impl Default for Document {
+	fn default() -> Self {
+		Self {
+			svg: SvgElement::Folder(Folder::default()),
+		}
+	}
 }
 
 impl Document {
 	pub fn render(&self) -> String {
-		self.svg.iter().map(|element| element.render()).collect::<Vec<_>>().join("\n")
+		self.svg.render()
+	}
+
+	pub fn open(&self, path: &str) -> Result<&SvgElement, DocumentError> {
+		assert!(matches!(self.svg, SvgElement::Folder(_)), "SVG root has to be of type Folder");
+		let mut root = &self.svg;
+		for s in path.split('/') {
```

Comment:
I'm thinking a vector (or slice) of strings would be safer than a slash-delimited path string. In case you want a slash in your path, for example, which could add complexity around escaping and room for mistakes.